dask/order.py

```diff
@@ -202,125 +198,286 @@ def dependencies_key(x):
+                        if len(finish_now) > 1:
+                            finish_now = sorted(finish_now, key=finish_now_key)
+                        for dep in finish_now:
+                            result[dep] = i
+                            i += 1
+                        add_to_inner_stack = False
+
+                if deps:
+                    for dep in deps:
+                        num_needed[dep] -= 1
+
+                    already_seen = deps & seen
+                    if already_seen:
+                        if len(deps) == len(already_seen):
+                            continue
+                        add_to_inner_stack = False
+                        deps -= already_seen
+
+                    if len(deps) == 1:
+                        # Fast path!  We trim down `deps` above hoping to reach here.
+                        (dep,) = deps
+                        if not inner_stack:
+                            if add_to_inner_stack:
+                                inner_stack = [dep]
+                                inner_stack_pop = inner_stack.pop
+                                seen_add(dep)
+                                continue
+                            key = partition_keys[dep]
+                        else:
+                            key = partition_keys[dep]
+                            if key < partition_keys[inner_stack[0]]:
+                                # Run before `inner_stack` (change tactical goal!)
+                                inner_stacks_append(inner_stack)
+                                inner_stack = [dep]
+                                inner_stack_pop = inner_stack.pop
+                                seen_add(dep)
+                                continue
+                        if key < partition_keys[item]:
+                            next_nodes[key].append(deps)
+                        else:
+                            later_nodes[key].append(deps)
+                    else:
+                        # Slow path :(.  This requires grouping by partition_key.
+                        dep_pools = defaultdict(list)
+                        for dep in deps:
+                            dep_pools[partition_keys[dep]].append(dep)
+                        item_key = partition_keys[item]
+                        if inner_stack:
+                            # If we have an inner_stack, we need to look for a "better" path
+                            prev_key = partition_keys[inner_stack[0]]
+                            now_keys = []  # < inner_stack[0]
+                            for key, vals in dep_pools.items():
+                                if key < prev_key:
+                                    now_keys.append(key)
+                                elif key < item_key:
+                                    next_nodes[key].append(vals)
+                                else:
+                                    later_nodes[key].append(vals)
```

Comment:
This isn't currently covered by testing.

Reply:
Thanks for giving this a look and a try @jcrist.  I know this code isn't exactly light reading.

I agree it would be nice to have tests that illustrate the usefulness of these code sections in `test_order.py`, but since these lines are covered by other dask tests, I don't consider such tests essential.  "Large and messy" graphs, which these lines need, are kind of difficult to make tests for.  Nevertheless, I'll try to make some and will let you know if it requires more effort than I'd like to give.

Reply:
No worries. I'm more concerned about the branch on line 439 than this one, since that branch is a fair chunk of code. Best effort within reason sounds good to me.

Reply:
I removed the branch at 439 after making a benchmark trying to exercise it: https://github.com/dask/dask-benchmarks/pull/40.  There was no performance difference.  This branch was hit more often in earlier versions of this PR, but it is no longer necessary (maybe it never was).  Good instincts @jcrist!